Tutorial Section 2 [2/3] (#548)
open-telemetry/opentelemetry-java-instrumentation: instrumentation-core/spring/README.md

       
### Project Background

> Here we will create RestControllers for `FirstService` and `SecondService`.

Reviewer: I think this reads better:
```suggestion
Here we will create REST controllers for `FirstService` and `SecondService`.
```

Required dependencies and configurations for FirstService and SecondService projects can be found [here](#setup-for-manual-instrumentation).

> ### Instrumentation of Receiving Service: FirstService

Reviewer: i think the names "Receiving Service" and "Client Service" are not too clear

i like the idea of trying to give them more descriptive names, instead of `FirstService` and `SecondService` though

maybe `MainService` / `TimeService`?

or `FrontService` / `TimeService` ?

Author: MainService and TimeService sounds good

@RestController
@RequestMapping(value = "/message")
public class FirstServiceController {
   private static int requestCount = 1;
   
   @Autowired
   private Tracer tracer;

   @Autowired
   private HttpUtils httpUtils;

>    private static final String SECOND_SERVICE_URL = "http://localhost:8081/time";

Reviewer: common for static fields to be at top of class

public class SecondServiceApplication {

   public static void main(String[] args) throws IOException {
      SpringApplication.run(SecondServiceApplication.class, args);
   }
}
```

Add the REST controller below to your SecondService project. This controller will return a string when SecondServiceController.secondTracedMethod is called:

```java
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping(value = "/time")
public class SecondServiceController {
   @Autowired
   private Tracer tracer;

   @GetMapping
   public String secondTracedMethod() {
      return "It's time to get a watch";
   }
}
```

#### Create Controller Filter

Add the class below to wrap all requests to the SecondServiceController in a span. This class will call the preHandle method before the REST controller is entered and the postHandle method after a response is created. 

The preHandle method starts a span for each request. This implementation is shown below:    

```java

@Component

> public class ControllerFilter implements Filter {

Reviewer: i forget, does this get automatically picked up by Spring, do we need to define a url mapping for it?

Run FirstService and SecondService from command line or using an IDE. The end point of interest for FirstService is `http://localhost:8080/message` and  `http://localhost:8081/time` for SecondService. Entering `localhost:8080/message` in a browser should call FirstService and then SecondService, creating a trace. 

***Note: The default port for the Apache Tomcat is 8080. On localhost both FirstService and SecondService services will attempt to run on this port raising an error. To avoid this add `server.port=8081` to the resources/application.properties file. Ensure the port specified corresponds to port referenced by FirstServiceController.SECOND_SERVICE_URL. ***

Congrats, we just created a distributed service with OpenTelemetry!

## Manual Instrumentation using Handlers and Interceptors

> In this section, we will implement the Spring HandlerInerceptor interface to wrap all requests to FirstService and Second Service controllers in a span. 

Reviewer: ```suggestion
In this section, we will implement the Spring HandlerInerceptor interface to wrap all requests to FirstService and SecondService controllers in a span. 
```

We will also use the RestTemplate HTTP client to send requests from FirstService to SecondService. To propagate the trace in this request we will also implement the ClientHttpRequestInterceptor interface. This implementation is only required for projects that send outbound requests. In this example it is only required for FirstService. 

> ### Setup FirstService and SecondService

Reviewer: two alternate suggestions here:

```suggestion
### Set up FirstService and SecondService
```

```suggestion
### Setup for FirstService and SecondService
```

> Using the earlier instructions [create two example projects](#create-two-spring-projects) and add the required [dependencies and configurations](#setup-for-manual-instrumentation). 

Reviewer: ```suggestion
Using the earlier instructions [create two spring projects](#create-two-spring-projects) and add the required [dependencies and configurations](#setup-for-manual-instrumentation). 
```

### Instrumentation of Client Service: SecondService

Ensure the main method in SecondServiceApplication is defined. This will be the entry point to the SecondService project. This file should be created by the Spring Boot project initializer.

```java
import java.io.IOException;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

@SpringBootApplication
public class SecondServiceApplication {

   public static void main(String[] args) throws IOException {
      SpringApplication.run(SecondServiceApplication.class, args);
   }
}
```

Add the RestController below to your SecondService project. This controller will return a string when SecondServiceController.secondTracedMethod is called:

```java
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

> import io.opentelemetry.context.Scope;
> import io.opentelemetry.trace.Span;
> import io.opentelemetry.trace.Tracer;
> 

Reviewer: these imports aren't used
```suggestion
```

@RestController
@RequestMapping(value = "/time")
public class SecondServiceController {
   @Autowired
   private Tracer tracer;

   @GetMapping
   public String secondTracedMethod() {
      return "It's time to get a watch";
   }
}
```

#### Create Controller Interceptor

Add the class below to wrap all requests to the SecondServiceController in a span. This class will call the preHandle method before the RestController is entered and the postHandle method after a response is created. 

> The preHandle method starts a span for each request. The postHandle method closes the span and adds the span context to the response header. This implementation is shown below:    

Reviewer: no need to add span context to the response header
```suggestion
The preHandle method starts a span for each request. This implementation is shown below:    
```

```java

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.web.servlet.HandlerInterceptor;
import org.springframework.web.servlet.ModelAndView;
import io.grpc.Context;
import io.opentelemetry.OpenTelemetry;
import io.opentelemetry.context.propagation.HttpTextFormat;
import io.opentelemetry.trace.Span;
import io.opentelemetry.trace.Tracer;
import io.opentelemetry.trace.TracingContextUtils;

@Component
public class ControllerTraceInterceptor implements HandlerInterceptor {

   private static final HttpTextFormat.Getter<HttpServletRequest> getter =
         new HttpTextFormat.Getter<HttpServletRequest>() {
            public String get(HttpServletRequest req, String key) {
               return req.getHeader(key);
            }
         };

>    private static final HttpTextFormat.Setter<HttpServletResponse> setter =
>          new HttpTextFormat.Setter<HttpServletResponse>() {
>             @Override
>             public void set(HttpServletResponse response, String key, String value) {
>                response.addHeader(key, value);
>             }
>          };
> 

Reviewer: no need to add span context to the response header
```suggestion
```

   @Autowired
   private Tracer tracer;

   @Override
   public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler)
         throws Exception {
     
      Span span = createSpanWithParent(request, context);
      span.setAttribute("handler", "pre");

>       tracer.withSpan(span);

Reviewer: it looks like this `Scope` is never closed?

Reviewer: this is important to address because the scope is bound to `ThreadLocal` and so will leak, and if/when that thread is reused it will have wrong/old span bound to it

Author: Interesting. I noticed that the HandlerInterceptor.postHandle is not called when an unhandled exception is thrown. So I moved this code to afterCompletion but that doesn't address this threading issue. Would you recommend implementing a Serverlet Filter instead of a HandlerInterceptor to wrap controllers in a span?

Reviewer: servlet filter sounds good 👍

      return true;
   }

   @Override
   public void postHandle(HttpServletRequest request, HttpServletResponse response, Object handler,
         ModelAndView modelAndView) throws Exception {

      Span currentSpan = tracer.getCurrentSpan();
      currentSpan.setAttribute("handler", "post");

>       OpenTelemetry.getPropagators().getHttpTextFormat().inject(Context.current(), response, setter);

Reviewer: no need to add span context to the response header
```suggestion
```

            @Override
            public void set(HttpServletResponse response, String key, String value) {
               response.addHeader(key, value);
            }
         };

   @Autowired
   private Tracer tracer;

   @Override
   public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler)
         throws Exception {
      Context context = OpenTelemetry.getPropagators().getHttpTextFormat()
          .extract(Context.current(), request, getter);
      Span span = createSpanWithParent(request, context);
      span.setAttribute("handler", "pre");
      tracer.withSpan(span);

      return true;
   }

   @Override
   public void postHandle(HttpServletRequest request, HttpServletResponse response, Object handler,
         ModelAndView modelAndView) throws Exception {

      Span currentSpan = tracer.getCurrentSpan();
      currentSpan.setAttribute("handler", "post");
      OpenTelemetry.getPropagators().getHttpTextFormat().inject(Context.current(), response, setter);
      currentSpan.end();
   }

   @Override
   public void afterCompletion(HttpServletRequest request, HttpServletResponse response,

>          Object handler, Exception exception) throws Exception {}

Reviewer: would it be better to end the span in `afterCompletion`?

Reviewer: bumping this, it seems github is hiding a couple of conversations even though they aren't marked resolved

### Instrumentation of Receiving Service: FirstService

Ensure the main method in FirstServiceApplication is defined. This will be the entry point to the FirstService project. This file should be created by the Spring Boot project initializer.

```java
@SpringBootApplication
public class FirstServiceApplication {

   public static void main(String[] args) throws IOException {
      SpringApplication.run(FirstServiceApplication.class, args);
   }
}
```

Create a RestController for FirstService. This controller will send a request to SecondService and then return the response to the client:

```java
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping(value = "/message")
public class FirstServiceController {
   @Autowired
   private Tracer tracer;
	
   @Autowired
   private RestTemplate restTemplate;

   @Autowired

>    HttpUtils httpUtils;

Reviewer: for consistency with the other fields
```suggestion
   private HttpUtils httpUtils;
```

>    private static String SECOND_SERVICE_URL = "http://localhost:8081/time";

Reviewer: constants are typically `static` and `final`
```suggestion
   private static final String SECOND_SERVICE_URL = "http://localhost:8081/time";
```

   @GetMapping
   public String firstTracedMethod() {

      ResponseEntity<String> response =
            restTemplate.exchange(SECOND_SERVICE_URL, HttpMethod.GET, null, String.class);
      String secondServiceTime = response.getBody();

      return "Second Service says: " + secondServiceTime;

   }
}
```

> As seen in the setup of SecondService, create implement the TraceInterceptor interface to wrap requests to the SecondServiceController in a span. Then register this new handler by extending the HandlerInterceptor. In effect, we will be taking a copy of the InterceptorConfig.java and ControllerTraceInterceptor.java defined in SecondService and adding it to FirstService. These files are referenced [here](#create-controller-interceptor).

Reviewer: ```suggestion
As seen in the setup of SecondService, implement the TraceInterceptor interface to wrap requests to the SecondServiceController in a span. Then register this new handler by extending the HandlerInterceptor. In effect, we will be taking a copy of the InterceptorConfig.java and ControllerTraceInterceptor.java defined in SecondService and adding it to FirstService. These files are referenced [here](#create-controller-interceptor).
```

#### Create Client Http Request Interceptor

Next, we will configure the ClientHttpRequestInterceptor to intercept all client HTTP requests made using RestTemplate.

> To propagate the span context from FirstService to SecondService we must inject the trace id and trace state into the outgoing request header. In section 1 this was done using the helper class HttpUtils. In this section, we will implement the ClientHttpRequestInterceptor interface and register this interceptor in our application. 

Reviewer: ```suggestion
To propagate the span context from FirstService to SecondService we must inject the trace parent and trace state into the outgoing request header. In section 1 this was done using the helper class HttpUtils. In this section, we will implement the ClientHttpRequestInterceptor interface and register this interceptor in our application. 
```

```java

import java.io.IOException;

import io.grpc.Context;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpRequest;

import org.springframework.http.client.ClientHttpRequestExecution;
import org.springframework.http.client.ClientHttpRequestInterceptor;
import org.springframework.http.client.ClientHttpResponse;
import org.springframework.stereotype.Component;
import io.opentelemetry.OpenTelemetry;
import io.opentelemetry.context.propagation.HttpTextFormat;
import io.opentelemetry.trace.Span;
import io.opentelemetry.trace.Tracer;

@Component
public class RestTemplateHeaderModifierInterceptor implements ClientHttpRequestInterceptor {

   @Autowired
   private Tracer tracer;

   private static final HttpTextFormat.Setter<HttpRequest> setter =
         new HttpTextFormat.Setter<HttpRequest>() {
            @Override
            public void set(HttpRequest carrier, String key, String value) {
               carrier.getHeaders().set(key, value);
            }
         };

   @Override
   public ClientHttpResponse intercept(HttpRequest request, byte[] body,
         ClientHttpRequestExecution execution) throws IOException {

>       Span currentSpan = tracer.getCurrentSpan();

Reviewer: this span is not ended

Author: I can see why this can be a bit cryptic but I started the span in ControllerTraceInterceptor.preHandle (line 442) and I end the span in ControllerTraceInterceptor.postHandle (line 454). `currentSpan` is opened and closed by the Controller Interceptor. Here I wrap each request in one span and I add events to this span in other parts of the application. I do not create sub spans with in a rest controller. In the RestTemplate interceptor defined above, I get the controller span from the tracer and I add an event. Although I'm not sure if this approach follows best practices for tracing or if I'm missing something here. Please advise and thanks again for your help.

      currentSpan.setAttribute("client_http", "inject");
      currentSpan.addEvent("Request sent to SecondService");

      OpenTelemetry.getPropagators().getHttpTextFormat().inject(Context.current(), request, setter);

      ClientHttpResponse response = execution.execute(request, body);

      return response;
   }
}

```

```java
import java.util.ArrayList;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.client.ClientHttpRequestInterceptor;
import org.springframework.web.client.RestTemplate;

@Configuration
public class RestClientConfig {

   @Autowired
   RestTemplateHeaderModifierInterceptor restTemplateHeaderModifierInterceptor;

   @Bean
   public RestTemplate restTemplate() {
      RestTemplate restTemplate = new RestTemplate();

>       List<ClientHttpRequestInterceptor> interceptors = restTemplate.getInterceptors();
>       if (interceptors.isEmpty()) {
>          interceptors = new ArrayList<>();
>       }
>       interceptors.add(restTemplateHeaderModifierInterceptor);
>       restTemplate.setInterceptors(interceptors);

Reviewer: ```suggestion
      restTemplate.getInterceptors().add(restTemplateHeaderModifierInterceptor);
```

          return req.getHeader(key);
        }
      };

  @Override
  public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
      throws IOException, ServletException {
    LOG.info("start doFilter");
    
    HttpServletRequest req = (HttpServletRequest) request;
    Context context = OpenTelemetry.getPropagators().getHttpTextFormat()
        .extract(Context.current(), req, GETTER);
    Span currentSpan = createSpanWithParent(req, context);
    try {
      tracer.withSpan(currentSpan);
      currentSpan.addEvent("dofilter");
      chain.doFilter(req, response);
    }finally {
      LOG.info("end doFilter");
      currentSpan.end();
    }
    
  }
  
  private Span createSpanWithParent(HttpServletRequest request, Context context) {
    Span parentSpan = TracingContextUtils.getSpan(context);
    Span.Builder spanBuilder = tracer.spanBuilder(request.getRequestURI()).setSpanKind(Span.Kind.SERVER);
    
    if (parentSpan.getContext().isValid()) {
      return spanBuilder.setParent(parentSpan).startSpan();
    }

    Span span = spanBuilder.startSpan();
    span.addEvent("Parent Span Not Found");

    return span;
  }
}
}

> 

Author: Good catch. It's an extra bracket. I removed it

   @Override
   public ClientHttpResponse intercept(HttpRequest request, byte[] body,
         ClientHttpRequestExecution execution) throws IOException {
      
      String spanName = request.getMethodValue() +  " " + request.getURI().toString();
      Span currentSpan = tracer.spanBuilder(spanName).setSpanKind(Span.Kind.CLIENT).startSpan();
      
      try {

>         OpenTelemetry.getPropagators().getHttpTextFormat().inject(Context.current(), request, setter);

Author: I added tracer.withSpan(currentSpan) to the try block
